### ISSUE TITLE
Sign packages with a strong name

### ISSUE THREAD

**Issue by johannesvp** [elmish/Elmish.WPF#308]
Our desktop application uses encryption to protect our IP. The encryptor requires the main assembly to be signed with a strong name. 
Visual Studio then requires all referenced assemblies to be strong named too. 

So the request is to strong name the deployed packages, Elmish and Elmish.Wpf. 

I think it is a simple change, that allows Elmish to be used by more developers.

**Comment by cmeeren** [comment]
I have never used strong naming myself. I have head various remarks about strong naming over the years, most of them negative. Allegedly, most of the warnings seem to apply to .NET Framework, not .NET Core or .NET 5+.

In any case, there is a project called [StrongNamer](https://github.com/dsplaisted/strongnamer) where anyone who needs strong-named packages just install StrongNamer as a normal NuGet package, and you don't have to worry about strong naming ever again. It seems that this is a much better solution:

* It allows devs like you to use whatever package they want, whether strong named or not. There's no need for devs to ask all the packages they need to implement strong naming.
* It's not "contagious". If we were to strong-name our NuGet, then that would require any packages we reference (such as [Elmish](https://github.com/elmish/elmish)) to also be strong-named, so the churn could spread across the ecosystem. (And AFAIK we would not be able to use StrongNamer to work around this, because devs would still pull Elmish and other references from NuGet, requring them to be strong-named, too.)

TL;DR: As long as a solution such as [StrongNamer](https://github.com/dsplaisted/strongnamer) works, I think that is a far better solution for all involved parts – the maintainers of Elmish.WPF (we don't have to implement SN or ask dependencies to do the same), the maintainers of our dependencies (ditto), and the devs who use Elmish.WPF (who don't have to ask for all their dependencies to be strong-named).

**Comment by TysonMN** [comment]
I agree with @cmeeren; `StrongNamer` seems like a good solution.

In my experience, most build servers obtain NuGet packages from `nuget.org`.  That site has a certificate, which should protect the build server from a man-in-the-middle attack.

However, if you want even more confidence that your packages are what you expect, then [check your packages into your repository](https://blog.ploeh.dk/2014/01/29/nuget-package-restore-considered-harmful/) after manually verifying that they are the expected ones.

**Comment by BentTranberg** [comment]
I worked on a project using strong naming from 2010 to 2016. It caused so much hassle in development that we ripped it out about midway. Don't remember the exact issues, so don't ask me. Except one reason to ditch it was that we wanted to use libraries that had no strong version, nor source we could build strong. From what I remember, the libraries we used came in strong and weak(?) versions, so I assume you shouldn't make a nuget library strong only. But as for me, never again!

**Comment by cmeeren** [comment]
> From what I remember, the libraries we used came in strong and weak(?) versions, so I assume you shouldn't make a nuget library strong only.

That is an anti-pattern. Acording to the official docs on [Strong naming and .NET Libraries](https://docs.microsoft.com/en-us/dotnet/standard/assembly/strong-named):

> ❌ DO NOT publish strong-named and non-strong-named versions of your library. For example, `Contoso.Api` and `Contoso.Api.StrongNamed`.

---

Also, I notice that the official docs on [Strong-named assemblies](https://docs.microsoft.com/en-us/dotnet/standard/assembly/strong-named) says:

> For .NET Core, strong-named assemblies do not provide material benefits.

This makes me even more certain that this is not worth the hassle. We support .NET Framework, but especially now that .NET 5 is out, fewer and fewer people will use .NET Framework, and the StrongNamer workaround exists for those who need it.

Closing this as I consider it resolved, but feel free to continue the discussion if you find it fruitful.

**Comment by BentTranberg** [comment]
> DO NOT publish strong-named and non-strong-named versions

I believe the meaning is "within one's environment", which makes sense, but not sure. I don't think they were thinking about nuget packages and libraries you can buy when they wrote that.
